File: frontend/src/types/coaching.ts
/**
 * Mirrors backend/app/schemas/coaching.py exactly.
 *
 * IMPORTANT (confirmed by reading app/api/v1/coaching.py):
 *   - GET  /coaching/profile                    -> CoachingProfile
 *   - GET  /coaching/insights                    -> CoachingInsights
 *   - GET  /coaching/recommendation              -> NextInterviewRecommendation
 *   - GET  /coaching/plans/active                -> PracticePlanResponse | null
 *   - POST /coaching/plans                       -> PracticePlanResponse (201)
 *   - POST /coaching/plans/{planId}/regenerate   -> PracticePlanResponse
 *   - PATCH /coaching/plans/{planId}/items/{itemId} -> PracticePlanResponse
 *
 * All routers are mounted directly on the app (no /api/v1 prefix), same
 * as /interview, /reports, /voice, /coding-interviews, /sessions.
 */

export type ReadinessLevel = "getting_started" | "building_confidence" | "interview_ready" | "highly_prepared";
export type CoachingInterviewType = "text" | "voice" | "coding";
export type PlanDuration = 7 | 14 | 30;
export type CoachingDifficulty = "easy" | "medium" | "hard";

export interface SessionTypeBreakdown {
  type: CoachingInterviewType;
  total: number;
  completed: number;
  in_progress: number;
  average_score: number;
  best_score: number;
}

export interface CoachingProfile {
  is_new_user: boolean;
  has_resume: boolean;
  target_role: string | null;
  total_sessions: number;
  completed_sessions: number;
  in_progress_sessions: number;
  average_score_overall: number;
  average_technical_score: number;
  average_communication_score: number;
  current_streak_days: number;
  longest_streak_days: number;
  strongest_topics: string[];
  weakest_topics: string[];
  by_type: SessionTypeBreakdown[];
}

export interface CoachingInsights {
  readiness_level: ReadinessLevel;
  readiness_score: number;
  top_strengths: string[];
  priority_weaknesses: string[];
  recommended_topics: string[];
  consistency_insight: string;
  next_milestone: string;
  ai_summary: string | null;
  generated_by_ai: boolean;
}

export interface NextInterviewRecommendation {
  interview_type: CoachingInterviewType;
  topic: string;
  difficulty: CoachingDifficulty;
  target_skill: string;
  reason: string;
  suggested_duration_minutes: number;
  suggested_number_of_questions: number;
  setup_path: string;
}

export interface PracticePlanItem {
  item_id: string;
  day_number: number;
  focus_area: string;
  interview_type: CoachingInterviewType;
  topic: string;
  difficulty: CoachingDifficulty;
  estimated_minutes: number;
  objective: string;
  recommended_activity: string;
  reason: string | null;
  completed: boolean;
}

export interface PracticePlanResponse {
  id: string;
  duration_days: PlanDuration;
  items: PracticePlanItem[];
  completed_items: number;
  total_items: number;
  completion_percentage: number;
  created_at: string;
}

export interface CreatePlanRequest {
  duration_days: PlanDuration;
}

export interface PlanItemUpdateRequest {
  completed: boolean;
}

export const PLAN_DURATIONS: PlanDuration[] = [7, 14, 30];

export const READINESS_LABELS: Record<ReadinessLevel, string> = {
  getting_started: "Getting Started",
  building_confidence: "Building Confidence",
  interview_ready: "Interview Ready",
  highly_prepared: "Highly Prepared",
};

File: backend/app/schemas/coaching.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.session_management import SessionTypeBreakdown

ReadinessLevel = Literal["getting_started", "building_confidence", "interview_ready", "highly_prepared"]
InterviewType = Literal["text", "voice", "coding"]
PlanDuration = Literal[7, 14, 30]


class CoachingProfile(BaseModel):
    """Snapshot of a user's practice history, built entirely from data the
    project already has (unified analytics + resume + interview history) -
    no duplicate score/streak computation, per the Module 9 requirements."""

    is_new_user: bool
    has_resume: bool
    target_role: str | None = None
    total_sessions: int
    completed_sessions: int
    in_progress_sessions: int
    average_score_overall: float
    average_technical_score: float
    average_communication_score: float
    current_streak_days: int
    longest_streak_days: int
    strongest_topics: list[str] = Field(default_factory=list)
    weakest_topics: list[str] = Field(default_factory=list)
    by_type: list[SessionTypeBreakdown] = Field(default_factory=list)


class CoachingInsights(BaseModel):
    readiness_level: ReadinessLevel
    readiness_score: float = Field(ge=0, le=100)
    top_strengths: list[str] = Field(default_factory=list)
    priority_weaknesses: list[str] = Field(default_factory=list)
    recommended_topics: list[str] = Field(default_factory=list)
    consistency_insight: str
    next_milestone: str
    ai_summary: str | None = None
    generated_by_ai: bool = False


class NextInterviewRecommendation(BaseModel):
    interview_type: InterviewType
    topic: str
    difficulty: Literal["easy", "medium", "hard"]
    target_skill: str
    reason: str
    suggested_duration_minutes: int
    suggested_number_of_questions: int
    setup_path: str


class PracticePlanItem(BaseModel):
    model_config = ConfigDict(extra="ignore")

    item_id: str
    day_number: int
    focus_area: str
    interview_type: InterviewType
    topic: str
    difficulty: Literal["easy", "medium", "hard"]
    estimated_minutes: int
    objective: str
    recommended_activity: str
    reason: str | None = None
    completed: bool = False


class PracticePlan(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str | None = None
    user_id: str
    duration_days: PlanDuration
    items: list[PracticePlanItem]
    is_active: bool = True
    created_at: datetime = Field(default_factory=datetime.utcnow)

    @property
    def completion_percentage(self) -> float:
        if not self.items:
            return 0.0
        done = sum(1 for item in self.items if item.completed)
        return round((done / len(self.items)) * 100, 1)


class PracticePlanResponse(BaseModel):
    id: str
    duration_days: PlanDuration
    items: list[PracticePlanItem]
    completed_items: int
    total_items: int
    completion_percentage: float
    created_at: datetime


class CreatePlanRequest(BaseModel):
    duration_days: PlanDuration = 7


class PlanItemUpdateRequest(BaseModel):
    completed: bool
